File: morphac/planners/base/test/planner_test.cc
#include "Eigen/Dense"
#include "gtest/gtest.h"

#include "constructs/include/trajectory.h"
#include "planners/include/planner.h"

namespace {

using Eigen::MatrixXd;

using morphac::constructs::Trajectory;
using morphac::planners::Planner;

// Derived Planner class for testing.
class CustomPlanner : public Planner {
 public:
  CustomPlanner(MatrixXd trajectory_data, int pose_size, int velocity_size)
      : Planner(), trajectory_(trajectory_data, pose_size, velocity_size) {}

  // ControlInput computation that always returns the same control input.
  Trajectory Compute() { return trajectory_; }

 private:
  Trajectory trajectory_;
};

class PlannerTest : public ::testing::Test {
 protected:
  PlannerTest() {
    // Set random seed for Eigen.
    srand(7);
  }

  void SetUp() override {}
};

TEST_F(PlannerTest, Construction) {
  MatrixXd trajectory_data1 = MatrixXd::Random(1, 5);
  MatrixXd trajectory_data2 = MatrixXd::Random(60, 6);
  CustomPlanner planner1{trajectory_data1, 3, 2};
  CustomPlanner planner2{trajectory_data2, 2, 4};
}

TEST_F(PlannerTest, Compute) {
  MatrixXd trajectory_data1 = MatrixXd::Random(1, 5);
  MatrixXd trajectory_data2 = MatrixXd::Random(60, 6);
  CustomPlanner planner1{trajectory_data1, 3, 2};
  CustomPlanner planner2{trajectory_data2, 2, 4};

  ASSERT_TRUE(planner1.Compute().get_data().isApprox(trajectory_data1));
  ASSERT_TRUE(planner2.Compute().get_data().isApprox(trajectory_data2));
}

}  // namespace
int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

File: morphac/planners/base/python/test/planner_test.py
import numpy as np
import pytest

from morphac.constructs import Trajectory
from morphac.planners.base import Planner


class CustomPlanner(Planner):
    def __init__(self, trajectory_data, pose_size, velocity_size):

        Planner.__init__(self)
        self.trajectory_data = trajectory_data
        self.pose_size = pose_size
        self.velocity_size = velocity_size

    def compute(self):

        return Trajectory(
            self.trajectory_data, self.pose_size, self.velocity_size)


@pytest.fixture()
def generate_planner_list():
    np.random.seed(7)
    p1 = CustomPlanner(np.random.randn(1, 3), 2, 1)
    p2 = CustomPlanner(np.random.randn(20, 5), 2, 3)

    return p1, p2


def test_compute(generate_planner_list):

    p1, p2 = generate_planner_list

    np.random.seed(7)
    data1 = np.random.randn(1, 3)
    data2 = np.random.randn(20, 5)

    assert np.allclose(p1.compute().data, data1)
    assert np.allclose(p2.compute().data, data2)
